### ISSUE TITLE
Popup position not correct when offsetWidth/offsetHeight == 0

### ISSUE THREAD

**Issue by jon48** [popperjs/popper-core#1351]
### CodeSandbox demo

https://codesandbox.io/s/gifted-sunset-ewyjj

### Steps to reproduce the problem

1. Using Chrome v92.0.4515.159
2. Make sure that popperjs/core is in version 2.9.3
3. Click on the red box, which is the trigger for displaying the pop-up

(The example is based on a largely more complex real use case, using Bootstrap, but I managed to recreate the behavior with that simple structure and pure Popper)

### What is the expected behavior?

The popup should be displayed just below the red box, but is instead translated to the top left, partly outside of the container.
The behaviour is correct in Firefox 91.0.2.

![image](https://user-images.githubusercontent.com/5150782/131545609-5a792176-cedc-4375-80cd-5fb8df68bbdb.png)

### What went wrong?

The issue was introduced by fix #1247, which introduce additional checks on scaled elements.

There are 2 main characteristics in the constructed scenario that triggers the problem:

- The parent container has non-integer margins (because of the use of fraction of rems: `margin: 1.1rem;`) that results in a `getBoundingClientRect` rectangle with decimal width and/or height. On the contrary, `offsetWidth` and `offsetHeight` return integers, so as the result the new `isElementScaled` function return `true`, as the numbers are not exactly the same.
  For instance, in my example, `rect.width == 321.859375`, and `element.offsetWidth == 321`, hence  `rect.width / element.offsetWidth || 1 == 1.0026771806853583`, and `scaleX !== 1` return `true`

- Independently, it appears that Chrome does not compute any offsetWidth or offsetHeight for the `#popper-button` element in this scenario, and return 0 for both of them (this is the difference with Firefox, as the latter return values).

Putting those two characteristics together in 

https://github.com/popperjs/popper-core/blob/f4a6550be3bc95ea5a0348218ae470848ff91544/src/dom-utils/getBoundingClientRect.js#L15-L19

`scaleX` and `scaleY` are computed because `includeScale == true`, and they are equal to `Infinity` because `element.offsetWidth == 0` and `element.offsetWidth == 0`

Then all values in the object returned by `getBoundingClientRect` are equals to 0 because they are divided by `Infinity`

So, when used in `getCompositeRect` function, with `const rect = getBoundingClientRect(...)`

https://github.com/popperjs/popper-core/blob/f4a6550be3bc95ea5a0348218ae470848ff91544/src/dom-utils/getCompositeRect.js#L56-L61

The `rect` components in the `x` and `y` output are 0, and - putting the scroll contribution aside - only the parent container offset negatively contributes, so the position is likely to be negative, therefore the translation to the top left corner.

#### Possible fixes

I am not familiar enough with the library to suggest the best course of action to prevent this bug, but I can see 2 directions:

- `isElementScaled` should maybe only compare integers, and round the `getBoundingClientRect` dimensions to integers. I guess that the initial fix was for significant scaling, not just when the bounding rectangle and the elements differs by a fraction of a pixel.
- Ignore scaling when `element.offsetWidth` or `element.offsetHeight` return 0, and prevent the `Infinity` scaling ratio.

### Any other comments?

Sorry if the bug description lacks more understanding of the library, I am not familiar with it, as I use it only through Bootstrap. I just observed the reported behaviour and concluded this was down to Popper through Javascript troubleshooting in Chrome Dev Tools.

**Comment by BrianHung** [comment]
> The parent container has non-integer margins (because of the use of fraction of rems: margin: 1.1rem;) that results in a getBoundingClientRect rectangle with decimal width and/or height. On the contrary, offsetWidth and offsetHeight return integers, so as the result the new isElementScaled function return true, as the numbers are not exactly the same.

Having a near exact same problem as this, but its parent container having non-integer width and height.

`40.5px` width and height
![40 5](https://user-images.githubusercontent.com/15721634/131590031-182abef3-ec68-4892-a852-316e82e6fffe.PNG)

`40.0px` width and height
![40](https://user-images.githubusercontent.com/15721634/131590034-0c1ee669-ba06-4a61-969a-85041a440747.PNG)

**Comment by FutyWidget** [comment]
I'm having the same problemen with version 2.9.3. With version 2.9.2 I experience no problems.

**Comment by FezVrasta** [comment]
May someone verify if this fixes your issue?

https://github.com/popperjs/popper-core/pull/1353

**Comment by Slapbox** [comment]
@FezVrasta I tried testing this by setting `@popperjs/core` to point at `"git+https://github.com/popperjs/popper-core.git#14c40f031872510b792c2239c16388500f0a4cea"` but I hit this issue when installing:

```
[4/4] Building fresh packages...
error C:\projects\app\node_modules\@popperjs\core: Command failed.
Exit code: 1
Command: husky install .config/husky
Arguments:
Directory: C:\projects\app\node_modules\@popperjs\core
Output:
'husky' is not recognized as an internal or external command,
```

So I installed `husky` but I just get more problems:

```
node_modules\husky\lib\commands\install.js:20
        throw new Error(".git can't be found");
```

Can you share the `dist` folder so I can drop it in place without using `yarn`?

**Comment by FezVrasta** [comment]
[Archive.zip](https://github.com/popperjs/popper-core/files/7096676/Archive.zip)

**Comment by jon48** [comment]
> May someone verify if this fixes your issue?

It does for me, in my initial use case. The dropdown is now correctly positioned.

**Comment by Slapbox** [comment]
Confirming this works for us as well. Thank you @FezVrasta!

**Comment by Slapbox** [comment]
Agh I think I'm mistaken. In our app we user Popper on anchor tags that can be just one line or can wrap to two lines. It looks like one-line anchor tags are still broken but I haven't any additional details to report yet.

**Comment by BrianHung** [comment]
Still broken for me (using Popper on a button element indirectly through `tippy/react`).

**Comment by Slapbox** [comment]
Since my last comment it seems to be working again but no changes to our `yarn.lock` - so maybe something weird happened. For now, I'm back in the "fixed" column. 

@BrianHung can you provide any more details?

**Comment by Slapbox** [comment]
And as soon as I've said that, I'm noticing that our Popper is moving up the Y axis relative to the target as the container is scrolled downwards. I think that explains the inconsistencies in my findings. There's definitely a difference in behavior.

**Comment by BrianHung** [comment]
> @BrianHung can you provide any more details?

I can provide some source code:

```jsx
<Tippy disabled={open} content={<span className="bg-gray-900 text-white py-1 px-2 rounded-md sm:rounded text-sm h-7 inline-block">Share</span>} duration={[25, 0]} zIndex={10} appendTo="parent">
  <Menu.Button className="h-9 w-9 rounded-full text-black dark:text-white bg-white bg-opacity-75 hover:bg-gray-200 focus:bg-gray-300 active:bg-gray-300 transition duration-[25ms]" aria-label="share">
    <svg className="h-5 w-5 m-auto fill-current">
      <use href={`${spritePath}#people-duo`}/>
    </svg>
  </Menu.Button>
</Tippy>
```

`h-9` and `w-9` both resolve to `40.5px`. Same problem I mentioned [above](https://github.com/popperjs/popper-core/issues/1351#issuecomment-909736815).

**Comment by mzgajner** [comment]
Can confirm same issue as @BrianHung (width resolving to .5 px), still present in latest version (2.10.1 at the time of this comment), fixed by reverting to 2.9.2.

**Comment by FezVrasta** [comment]
The provided repro works with the latest Popper version, please open a new ticket if you still experience issues.


https://codesandbox.io/s/sad-wind-jbmqs
